ankimorphs/text_highlighting.py

```diff
@@ -1,238 +1,386 @@
-                # note: it can seem unnecessarily complicated to append
-                # the ruby character in this else branch, but since the
-                # if branch above can potentially also trigger on the
-                # subsequent index _and_ that takes priority, it means that
-                # this next ruby character might never be reached unless
-                # we do it here.
-                next_index = index + 1
-                if next_index in ruby_character_dict:
-                    non_span_string += ruby_character_dict[next_index]
-                    # this entry is not needed anymore
-                    del ruby_character_dict[next_index]
-
-            highlighted_text_list.append(non_span_string)
-            index += 1
-
-    # print(f'highlighted text: {"".join(highlighted_text_list)}')
-    return "".join(highlighted_text_list)
-
-
-def _extract_ruby_characters_and_filter_string(
-    am_config: AnkiMorphsConfig, text_to_highlight: str
-) -> tuple[dict[int, str], str]:
-    ruby_character_dict: dict[int, str] = {}
-
-    # most users probably don't have ruby characters on their cards,
-    # so we only want to do all this extra work of extracting and replacing
-    # if they have activated the relevant pre-process option
-    if not am_config.preprocess_ignore_bracket_contents:
-        return ruby_character_dict, text_to_highlight
-
-    while True:
-        # matches first found, left to right
-        match: re.Match[str] | None = re.search(
-            text_preprocessing.square_brackets_regex, text_to_highlight
-        )
-        if match is None:
-            break
+    def prefix_len(self) -> int:
+        return len("<ruby>")
 
-        ruby_character_dict[match.start()] = match.group()
+    def open(self) -> str:
+        return "<ruby>"
 
-        # remove the found match from the string and repeat
-        text_to_highlight = (
-            text_to_highlight[: match.start()] + text_to_highlight[match.end() :]
-        )
+    def close(self) -> str:
+        return "</ruby>"
 
-    return ruby_character_dict, text_to_highlight
+    def rt(self) -> str:
+        return f"<rt>{self.ruby}</rt>"
 
+    def rt_offset(self) -> int:
+        return len(self.base) - 1
 
-def _extract_span_elements_and_filter_string(
-    am_config: AnkiMorphsConfig, card_morphs: list[Morpheme], text_to_highlight: str
-) -> tuple[list[SpanElement], str]:
-    span_elements: list[SpanElement] = []
+    def inject(self, target: str) -> str:
+        return target[: self.start] + str(self) + target[self.end :]
 
-    # To avoid formatting a smaller morph contained in a bigger morph, we reverse sort
-    # the morphs based on length and extract those first.
-    morphs_by_size = sorted(
-        card_morphs,
-        key=lambda _simple_morph: len(_simple_morph.inflection),
-        reverse=True,
-    )
+    def __str__(self) -> str:
+        return f"<ruby>{self.base}<rt>{self.ruby}</rt></ruby>"
```

Comment:
maintaining the square brackets instead of converting it to html would make the native filters work again

Reply:
I'll give this a shot.... but the built in filters can't be run after custom ones, so we'll need a codepath for bulk run on the raw data (where users would do {{furigana|kana|kanji:am-highlighted}}) and one where the `am-highlight` pipe converts to <ruby>.

Reply:
How about having 4 implementations of `RubyRange`?

### 1. Default

Maintain the square brackets, used by `am-highlighted` and `am-highlight:`, this means native filters can be used on `am-highlighted`

### 2. Furigana

Converts brackets into html (current implementation), used by `am-highlight-furigana:`

### 3. Kanji

Strips the brackets and its content, used by `am-highlight-kanji`

### 4. Kana

Strips the base, just leave the ruby text, used by `am-highlight-kana:`

Reply:
> How about having 4 implementations of `RubyRange`?
> 
> ### 1. Default
> Maintain the square brackets, used by `am-highlighted` and `am-highlight:`, this means native filters can be used on `am-highlighted`
> 
> ### 2. Furigana
> Converts brackets into html (current implementation), used by `am-highlight-furigana:`
> 
> ### 3. Kanji
> Strips the brackets and its content, used by `am-highlight-kanji`
> 
> ### 4. Kana
> Strips the base, just leave the ruby text, used by `am-highlight-kana:`

Great minds think alike. yesterday, I tired this idea (...well, a simplified version) and subclassed a `RubyTextRange` with the same interface as RubyRange. The implementation worked, but the reason it won't quite work in practice is that we end up with output that looks like : `<span morph-status="unkown"...>本</span>[ほん]` and that can't be parsed by `furigana`, etc. (it will only grab up to the space in the span :( ) 

The only solve I came up with for the "default" version would be to try to make the output much like your original code. This would have the same downside though, because as you noted in the docs, unless we move all the way to html the furigana issue is not solvable. :(

I do like this idea for the other filters: `am-highlight-furigana:` `am-highlight-kanji` `am-highlight-kana:`.

Reply:
I just tried this quickly and sloppily:

```python
class RubyRange(Range):
    """Represents a ruby and its range in parent string."""

    def __init__(self, start: int, end: int, base: str, ruby: str):
        super().__init__(start, end)
        self.base = base
        self.ruby = ruby

    def prefix_len(self) -> int:
        return len("")

    def open(self) -> str:
        return ""

    def close(self) -> str:
        return ""

    def rt(self) -> str:
        return ""

    def rt_offset(self) -> int:
        return 0

    def inject(self, target: str) -> str:
        return target[: self.start] + str(self) + target[self.end :]

    def __str__(self) -> str:
        return f"{self.base}[{self.ruby}]"

    def __repr__(self) -> str:
        return f"Range: {self.start}-{self.end}. Value: {self.base}[{self.ruby}]."

```

which works on these:

```
text: 本[ほん]
highlighted text: <span morph-status="unknown">本[ほん]</span>

text: 百舌鳥[もず] 明日[あした]
highlighted text: <span morph-status="unknown">百舌鳥[もず]</span><span morph-status="unknown">明日[あした]</span>
```

it fails on this one though:
```
text: 錬金術師[れんきんじゅつし]
highlighted text: <span morph-status="unknown">錬金術</span><span morph-status="unknown">師</span>[れんきんじゅつし]
```

Reply:
yea, that's the case I was talking about. I cannot think of a way that we could format this case that the built-in processors would be able to parse, without reintroducing the former issue that I was solving by using full html rubies. -- that would involve moving the bracketed text into the prior <span>.

Some alternate solutions:
1. We could have dual paths, one for the pre-processing and one for the JIT processing (and potentially add all three of `am-highlight-furigana:`, `am-highlight-kanji:`, and `am-highlight-kana:`. pre-processing would suffer from the ruby misalignment issue still :(
2. I hate adding more settings, but we could also possibly ask the user what they prefer for the pre-processing step. then we use that setting to run the pre-processor and produce the desired output (both, kanji only, kana only).
3. Scrap the unified processing and revert this PR to be just JIT processing (that would have slightly different output than the pre-processing. :(
4. Scrap entirely; I know this is taking a lot of your time and energy and that wasn't my intent, especially as you've already indicated that you want to get some of your time back re: this project :(
5. Add `am-highlight-furigana:`, `am-highlight-kanji:`, and `am-highlight-kana:` and accept the pr. (possibly adding css workarounds as discussed before.
6. ??? I am all ears if you have other options.